### ISSUE TITLE
Pre-computation of integrals for Chebyshev projection

### ISSUE THREAD

**Issue by chenwilliam77** [RJDennis/SolveDSGE.jl#23]
I am confused by how this package pre-computes integrals for Chebyshev projection. The code snippet below pre-computes integrals in the package:
```
function compute_chebyshev_integrals(eps_nodes,eps_weights,order,sigma)

  integrals = Array{Float64}(undef,order+1)
  for i = 1:(order+1)
    integrals[i] = sum(exp.(sqrt(2)*sigma*(i-1)*eps_nodes).*eps_weights)*pi^(-1/2)
  end

  return integrals

end
```
My understanding is that this code calculates the Gauss-Hermite quadrature of `exp(i * eps)`, where `i` is the order of the polynomial approximation. The code then uses these integrals to re-scale the weights on each basis function, as shown below (Lines 733-744 of solution_functions.jl)
```
        for i = 1:length(jumps_approximated)
            weights[i] = chebyshev_weights(variables[jumps_approximated[i]],grid,order,domain)
        end

        scaled_weights = copy(weights)
        for i = 1:length(jumps_approximated)
            for j = 1:ns
                index = [1:ndims(weights[i]);]
                index[1],index[j] = index[j],index[1]
                scaled_weights[i] = permutedims(integrals[j].*permutedims(weights[i],index),index)
            end
        end
```
I double checked the code for `chebyshev_weights`, and these weights appear to be the coefficients on the order `i` Chebyshev polynomial (please correct me if I'm wrong).

In the Judd, Maliar, and Maliar paper, these are the correct integrals to calculate _when the approximating polynomials are ordinary polynomials_. To precompute integrals for other polynomial families (and other function classes), the authors provide Proposition 1. By that proposition, I believe the correct integrals should actually be something like
```
function compute_chebyshev_integrals(eps_nodes,eps_weights,order,sigma,basis_functions)

  integrals = Array{Float64}(undef,order+1)
  for i = 1:(order+1)
    f = basis_functions[i] # basis_functions is a vector of each basis function, which corresponds to the maximal order of the Chebyshev basis functions
    integrals[i] = sum(f(sqrt(2)*sigma*eps_nodes).*eps_weights)*pi^(-1/2)
  end

  return integrals

end
```

Please let me know if I'm misunderstanding either your code or the paper.

Best,

William

**Comment by RJDennis** [comment]
Hi William,

Back when I coded that up I ran some tests on various models.  The results indicated that the solution using the pre-computation of integrals rountines gave results that were the same as the standard Gauss-Hermite quadrature to about 5-6 significant figures.  I also tried using the expression from section 5.1 of the paper and the integral terms that adjust the weights were numerically very similar to those using the expression in the code.  Essentially, the polynomial transformation of the nodes introduced the (i-1) multiple in the coded expression.  If there is a mistake in there, then it's effects are small and I haven't been able to find it.

It is unfortunate that the licence agreement of Judd's et al's code means that I cannot look at it and still have SolveDSGE on an MIT licence.
